Feat: MsSql support

Angelelz opened the pull request:
This PR aims to close #585.

The initial implementation and the first couple of tests are implemented. I will update this description with the status of the implementation.
Initially the only supported driver will be `node-mssql`.
There is still a lot to do:

- [x] Driver implementation
- [x] Core functionality
- [x] Initial basic CRUD API
- [x] Initial basic CRUD API integration tests
- [x] Type tests
- [x] RQB API
-  ~~`Merge` support (to do upsert type of functionality)~~
- [x] schema integration tests
- [x] custom types integration tests
- [x] prefixed integration tests
- [x] Replica tests
- [x] relational tests
- [x] Check (and support) most (all?) mssql data types
- [x] Add support for `$onUpdate`
- [x] Others? ⏳

Angelelz committed:
[All] Added types and new config option to column and column-builder classes

Angelelz committed:
[MySql] Added implementation for dialect

Angelelz committed:
[Pg] Added implementation for dialect

Angelelz committed:
[SQLite] Added implementation for dialect

Angelelz committed:
[All] Improved performance of update builder in the dialect

Angelelz committed:
[All] Fixed insert and update behavior

Angelelz committed:
[All] Added integration tests

Angelelz committed:
[All] Added requested changes: fix issue ignoring null values on update

Angelelz committed:
[MsSql] initial implementation

Angelelz committed:
[MsSql] deleted deprecated methods and functions and fixed types for columns

Angelelz committed:
[MsSql] Delete serial data type and added identity to number data types

Angelelz committed:
[MsSql] Added mssql and @types/mssql to integration tests package

Angelelz committed:
[MsSql] Removed intersectAll and exceptAll

Angelelz committed:
[MsSql] Added a way to not insert default on identity columns

Angelelz committed:
[MsSql][teporary] made json column be a text

Angelelz committed:
[MsSql] datetime columns don't need to be mapped to a Date object mssql does that for us

Angelelz committed:
[MsSql] fix offset and fetch in select and allow insert default values

Angelelz committed:
[MsSql] Fix save transaction syntax

Angelelz committed:
[MsSql] Added limit and offset to select

Angelelz committed:
[MsSql] deleted not supported onDuplicateKey

Angelelz committed:
[MsSql] Removed unsopported ignore in insert

Angelelz committed:
[MsSql] Fixed syntax in migrate function

Angelelz committed:
[MsSql] Fixed integer columns and deleted not used types

Angelelz committed:
[MsSql] Created but data type

Angelelz committed:
[MsSql] Fixed string data types

Angelelz committed:
[MsSql] fixed time data types

Angelelz committed:
[MsSql] Added initial batch of integration tests

AndriiSherman commented:
@Angelelz I guess we should not include RQB in first iteration for MSSQL. So it can be another PR scope, same as `Merge support (to do upsert type of functionality)`

Angelelz committed:
[MsSql] Made config for char and varchar optional

Angelelz committed:
[MsSql] Added dialect support for top and for clauses in the select

Angelelz committed:
[MsSql] [feat] Added RQB support

Angelelz committed:
[MsSql] Added schema and RQB tests

Angelelz committed:
possible tests fix

Angelelz committed:
Revert "possible tests fix"

This reverts commit b33879e7e27b922902362be88caa88068de082fe.

Angelelz commented:
No idea why the tests are not passing. They run will locally. Any idea @dankochetov ?

Angelelz committed:
[MsSql] Added support for `$onUpdate` and added tests

Angelelz committed:
[MsSql] Allmost all columns and column features supported

Angelelz committed:
fixed integration tests types

Angelelz committed:
[MsSql] Added types tests

Angelelz committed:
[All] Added types for generated columns and fix insert and update types with identity in MsSql

Angelelz committed:
[All] Fixed types tests with the new generated columns types

Angelelz committed:
[MsSql] Added schema tests

Angelelz committed:
[MsSql] Added prefixed tests

Angelelz committed:
[MsSql] Added tests for custom types

Angelelz committed:
[MsSql] Added tests for replicas and fixed types

Angelelz commented:
Some features have been postponed to later PRs, the scope of this one is already huge as it is.
Here are some notable ones:
 - `Merge` clause (to do upserts)
 - `Output` clause (similar to returning)
 - `Select top` similar to limit
 - Recursive CTE ergonomics

Angelelz committed:
[MsSql] Improve the types for db.execute. It properly returns the type passed as generic parameter

hari-bhandari commented:
Hello guys, our team would greatly appreciate it if there is any update re this issue. Prisma seems to be too abstract, so we are looking to migrate in a few days. Do you guys have any updates on this, please?

Aaqu commented:
> Hello guys, our team would greatly appreciate it if there is any update re this issue. Prisma seems to be too abstract, so we are looking to migrate in a few days. Do you guys have any updates on this, please? @Angelelz

I think we should ask @AndriiSherman

AndriiSherman commented:
> > Hello guys, our team would greatly appreciate it if there is any update re this issue. Prisma seems to be too abstract, so we are looking to migrate in a few days. Do you guys have any updates on this, please? @Angelelz
> 
> I think we should ask @AndriiSherman

It's in the review phase. I hope that it will be shipped this month. Everything depends on the review process and whether there will be any major problems with this PR

Angelelz committed:
[MsSql] fix: use schema name in select when the table is aliased

andreasbhansen commented:
Has the preview phase month uncovered any issues potentially causing delay merging this feature? Very stoked about MSSQL support in Drizzle!

bitaccesscomau commented on drizzle-orm/src/column-builder.ts:
```diff
@@ -1,5 +1,6 @@
 import { entityKind } from '~/entity.ts';
 import type { Column } from './column.ts';
+import type { MsSqlColumn } from './mssql-core/index.ts';
```
Looks like a duplicate import on this line.
Apologies not familiar with git and unable to lodge a change myself.

GustavoOS commented:
Does this PR add support to Drizzle-kit?

andreasbhansen commented:
@Angelelz What are your thoughts on ["Insert returning"](https://orm.drizzle.team/docs/insert#insert-returning) and ["Update with returning"](https://orm.drizzle.team/docs/update#update-with-returning), allowing for 
 `{INSERT_OR_UPDATE_STATEMENT}.returning({ insertedId: users.id });` similar to the Postgres/SQLite drivers?

Using the [`OUTPUT` clause](https://learn.microsoft.com/en-us/sql/t-sql/queries/output-clause-transact-sql?view=sql-server-ver16) would return any specified values.

```
CREATE TABLE Todos (
    ID INT IDENTITY(1,1) PRIMARY KEY,
    Description NVARCHAR(255) NOT NULL,
    IsCompleted BIT NOT NULL DEFAULT 0,
    CreatedAt DATETIME2 NOT NULL DEFAULT GETDATE(),
);

INSERT INTO [Todos]
([Description], [IsCompleted])
OUTPUT INSERTED.ID, INSERTED.Description
VALUES ('A description', 0);

UPDATE [Todos]
SET Description='Updated description'
OUTPUT INSERTED.ID, INSERTED.Description
WHERE ID=1
```
Result:
![image](https://github.com/drizzle-team/drizzle-orm/assets/1404793/647e7e85-53a7-438d-97fc-2b19b4f34ff9)

At the moment, I believe we would need to write INSERT/UPDATE statements in the sql`` template, but of course it's more convenient to use the built-in Drizzle insert/update mechanisms. 
```
await db.execute(sql`
    INSERT INTO [Todos] ([Description], [IsCompleted])
    OUTPUT INSERTED.ID, INSERTED.Description
    VALUES (${form.description}, ${form.isCompleted});
`);
```

Angelelz commented:
That's on the list of notable features that have been postponed due to the size of the PR. See my comment [here](https://github.com/drizzle-team/drizzle-orm/pull/1610#issuecomment-1851217492)

Aaqu commented:
> > > Hello guys, our team would greatly appreciate it if there is any update re this issue. Prisma seems to be too abstract, so we are looking to migrate in a few days. Do you guys have any updates on this, please? @Angelelz
> > 
> > 
> > I think we should ask @AndriiSherman
> 
> It's in the review phase. I hope that it will be shipped this month. Everything depends on the review process and whether there will be any major problems with this PR

Any news? 🥺

GustavoOS commented:
It seems this PR has conflicts. Do they have to be resolved before or after the review?

iaingymware commented:
When will this be good to go?  

Looking to use Drizzle as our ORM but all DB within the org are MSSQL in Azure.

phildaniels commented:
Can say that as soon as this gets merged, we are immediately migrating from prisma/sequelize at my org!

iaingymware commented:
Looks like this isn't going to get merged any time soon so will have to go with an alternative solution.  Unfortunate.

andreasbhansen commented:
> Looks like this isn't going to get merged any time soon so will have to go with an alternative solution. Unfortunate.

The issue is listed as in-progress: https://github.com/orgs/drizzle-team/projects/1?pane=issue&itemId=47747581

scottrupprecht commented:
I definitely get the never ending priorities for the team, is there anything we in the community can do to assist in moving this forward? Maybe just the orm functionality released in a beta branch so we can assist in the review while the drizzle kit work is pending? I have manually compiled this PR and tested it extensively over the past few months to much success. I'd have no problem putting money where my mouth is so I'm not asking for something for nothing in return, I'd be happy to bump up my monthly sponsorship or shooting over a one-time payment and I hope others would consider doing so as well to get this functionality into this fine code base.

lroal commented:
An intermediate solution would be to use [GitHub.com/alfateam/rdb](GitHub.com/alfateam/rdb)

It is extremely fast for mssql as it uses FOR JSON PATH operator

Aaqu commented:
Any news? 🥺 v2

GustavoOS commented:
> I definitely get the never ending priorities for the team, is there anything we in the community can do to assist in moving this forward? Maybe just the orm functionality released in a beta branch so we can assist in the review while the drizzle kit work is pending? I have manually compiled this PR and tested it extensively over the past few months to much success. I'd have no problem putting money where my mouth is so I'm not asking for something for nothing in return, I'd be happy to bump up my monthly sponsorship or shooting over a one-time payment and I hope others would consider doing so as well to get this functionality into this fine code base.

Maybe this is something you should take this subject into the discord channel: https://discord.gg/tCe773yFeZ

TG-DylanH commented:
My company is starting a new major project soon, and I'd love to use this over Prisma. However, we only use MSSQL. It looks like there hasn't been any progress in ~5 months, could we please get an update on whether this is going to be merged in the near future? 🥺

andreasbhansen commented:
Latest update afaik, found on Discord ([permalink](https://discord.com/channels/1043890932593987624/1255835097156812891/1255839467521048659)):
> **Andrew Sherman (06/27/2024 12:57 PM):**  will be done after RLS support, which will be done after generated columns(which I'm pushing to preview release in 1hr and sending discord notification about)

Aaqu commented:
RLS info https://github.com/drizzle-team/drizzle-orm/discussions/2450#discussioncomment-10491262
>  **Andrew Sherman (08/29/2024):** Update: finishing pull and push logic for policies and roles and preparing for beta release

itpropro commented:
> RLS info [#2450 (comment)](https://github.com/drizzle-team/drizzle-orm/discussions/2450#discussioncomment-10491262)
> 
> > **Andrew Sherman (08/29/2024):** Update: finishing pull and push logic for policies and roles and preparing for beta release

This is for PostgreSQL's Row Level Security (RLS), not for MSSQL or am I missing something?

itpropro commented:
> RLS info [#2450 (comment)](https://github.com/drizzle-team/drizzle-orm/discussions/2450#discussioncomment-10491262)
> 
> > **Andrew Sherman (08/29/2024):** Update: finishing pull and push logic for policies and roles and preparing for beta release

This is still not related to MSSQL, please stop referencing this thread.

andrewchester1 commented:
Any update on this feature, or a general timeline? The team I am working on is hoping to switch to a newer typescript api framework, but we need mssql support to make this switch. It would be amazing to use drizzle, and hope this feature comes out soon!

mihaipoenaru commented:
I am also starting a project for my company and it would be great if I could just start using drizzle as opposed to making stuff by hand and then going through the pain of porting everything to drizzle.

tmrclark commented:
For anybody else who is looking for a good mssql compatible ORM but can't wait for drizzle to add support, my org ended up using [kysely](https://kysely.dev/) and have been very happy with it. It has a companion code gen tool that can introspect types from the database, making setup straightforward.

Angelelz committed:
Merge branch 'master' into internal-mssql

Angelelz committed:
fix type and linting errors

Angelelz committed:
fixed all conflicts but having circular dependency issues

Angelelz committed:
fix circular deps and moved mssql test to vitest

Angelelz committed:
fixed type error in tests

Angelelz commented:
The PR has been updated to resolve conflicts and use the new structures/features of the library.
But there is still some work to do: notably, the new syntax to define columns without the database column name is not implemented. @AndriiSherman Should I implement that feature for this PR or you would prefer to include that feature at a later date?

ImBoop commented:
What else is needed to bring this to mainline? TypeORM has recently got a new breath of support, while Drizzle seems to be losing their momentum, is there anything we can do to help move it along?

Aaqu commented:
Tomorrow is the anniversary of the pull request . Will it be merged... 🥺 so we can celebrate 🎉🥳🎈?

Gideon-Felt commented:
Is no one going to respond to @Angelelz? next steps are needed.

Angelelz commented:
Just to put this out there. This is a 40k+ line PR, and this is only half of what is needed to support mssql. The team will do its best but remember that drizzle-kit support will be needed before this is merged. This PR is only for the ORM side. Mssql requires new syntax to introspect, generate migrations, support the studio, and other drizzle-kit stuff. My understanding is that the team will take care of that part.

opensas commented:
> Just to put this out there. This is a 40k+ line PR, and this is only half of what is needed to support mssql. The team will do its best but remember that drizzle-kit support will be needed before this is merged. This PR is only for the ORM side. Mssql requires new syntax to introspect, generate migrations, support the studio, and other drizzle-kit stuff. My understanding is that the team will take care of that part.

great, thanks a lot for all this effort

it would be great to have an issue with the remaining tasks, so we can track it, and it would help everybody understand what's still missing

AndriiSherman committed:
Merge branch 'beta' of github.com:drizzle-team/drizzle-orm into feat-mssql-support

AndriiSherman committed:
Fix build error

mfferreira commented:
Any progress on this?

ProvorovOleksii committed:
Updated:
- Added default handling for drizzle-kit
- Changed config type for mssql tables

AleksandrSherman committed:
merge main to feat-mssql-support

AleksandrSherman committed:
casing + data types + joins tests

Removed:
- smalldate and mediumint columns
Added:
- snake_case and camelCase to config
Updated:
- table config and some columns configs. No need to pass names to columns
- View config

AleksandrSherman committed:
Fixed type errors in test for mssql
Fixed dprint issues
Commented mssql dialect in drizzle-kit

AleksandrSherman committed:
WIP limit offset

AleksandrSherman committed:
WIP limit offset

AleksandrSherman committed:
[mssql] mandatory constraint names + tests

Updated:
- tests (added names to constraints)
- made constraint names mandatory for mssql
> Mssql generates names with custom hash
- top, offset and fetch are now accept placeholder

Added:
- drizzle mock
- new tests for top, offset and fetch

Also commented mssql logic in drizzle-kit as it is not supported yet

AleksandrSherman committed:
Merge branch 'main' of https://github.com/drizzle-team/drizzle-orm into feat-mssql-support

AleksandrSherman committed:
[mssql]: fixed tests

AleksandrSherman committed:
[mssql]: fixed deps

AleksandrSherman committed:
Merge branch 'main' of https://github.com/drizzle-team/drizzle-orm into feat-mssql-support

AleksandrSherman committed:
[mssql] constraint names are not mandatory now

Updated tests

AleksandrSherman committed:
[mssql] commented drizzle-kit logic

AleksandrSherman committed:
Merge branch 'main' of https://github.com/drizzle-team/drizzle-orm into feat-mssql-support

AleksandrSherman committed:
[mssql]: fixed imports

Gideon-Felt commented:
> great, thanks a lot for all this effort
> 
> it would be great to have an issue with the remaining tasks, so we can track it, and it would help everybody understand what's still missing
@Angelelz wondering if you saw this? I think they want more visibility on this, I know that would certainly get more eyes on it at least.

mfferreira commented:
Can't wait to use this!

Aaqu commented:
@AndriiSherman any news on mssql?

Aaqu commented:
There is a new version of mssql 12.0.0, will it be compatible?
